File: unidbg-android/src/main/java/com/vortexdbg/linux/file/NullFileIO.java
package com.vortexdbg.linux.file;

import com.vortexdbg.Emulator;
import com.vortexdbg.arm.backend.Backend;
import com.vortexdbg.file.FileIO;
import com.vortexdbg.file.linux.BaseAndroidFileIO;
import com.vortexdbg.file.linux.IOConstants;
import com.vortexdbg.file.linux.StatStructure;
import com.sun.jna.Pointer;

import java.io.IOException;
import java.util.Arrays;

public class NullFileIO extends BaseAndroidFileIO implements FileIO {

    private final String path;

    public NullFileIO(String path) {
        super(IOConstants.O_RDWR);

        this.path = path;
    }

    private boolean isTTY() {
        return "/dev/tty".equals(path);
    }

    @Override
    public void close() {
    }

    @Override
    public int write(byte[] data) {
        if (isTTY()) {
            try {
                System.out.write(data);
            } catch (IOException e) {
                throw new IllegalStateException(e);
            }
        }

        return data.length;
    }

    @Override
    public int lseek(int offset, int whence) {
        return 0;
    }

    @Override
    public int read(Backend backend, Pointer buffer, int count) {
        if (isTTY()) {
            try {
                byte[] buf = new byte[count];
                int read = System.in.read(buf);
                if (read <= 0) {
                    return read;
                }
                buffer.write(0, Arrays.copyOf(buf, read), 0, read);
                return read;
            } catch (IOException e) {
                throw new IllegalStateException(e);
            }
        }
        return 0;
    }

    @Override
    public int fstat(Emulator<?> emulator, StatStructure stat) {
        stat.st_size = 0;
        stat.st_blksize = 0;
        stat.pack();
        return 0;
    }

    @Override
    public int ioctl(Emulator<?> emulator, long request, long argp) {
        return 0;
    }

    @Override
    public String toString() {
        return path;
    }
}

File: vortexdbg-android/src/main/kotlin/com/vortexdbg/linux/file/NullFileIO.kt
package com.vortexdbg.linux.file

import com.vortexdbg.Emulator
import com.vortexdbg.arm.backend.Backend
import com.vortexdbg.file.FileIO
import com.vortexdbg.file.linux.BaseAndroidFileIO
import com.vortexdbg.file.linux.IOConstants
import com.vortexdbg.file.linux.StatStructure
import com.sun.jna.Pointer

import java.io.IOException
import java.util.Arrays

open class NullFileIO(private val path: String) : BaseAndroidFileIO(IOConstants.O_RDWR), FileIO {

    private fun isTTY(): Boolean {
        return "/dev/tty" == path
    }

    override fun close() {
    }

    override fun write(data: ByteArray): Int {
        if (isTTY()) {
            try {
                System.out.write(data)
            } catch (e: IOException) {
                throw IllegalStateException(e)
            }
        }

        return data.size
    }

    override fun lseek(offset: Int, whence: Int): Int {
        return 0
    }

    override fun read(backend: Backend, buffer: Pointer, count: Int): Int {
        if (isTTY()) {
            try {
                val buf = ByteArray(count)
                val read = System.`in`.read(buf)
                if (read <= 0) {
                    return read
                }
                buffer.write(0, Arrays.copyOf(buf, read), 0, read)
                return read
            } catch (e: IOException) {
                throw IllegalStateException(e)
            }
        }
        return 0
    }

    override fun fstat(emulator: Emulator<*>, stat: StatStructure): Int {
        stat.st_size = 0L
        stat.st_blksize = 0
        stat.pack()
        return 0
    }

    override fun ioctl(emulator: Emulator<*>, request: Long, argp: Long): Int {
        return 0
    }

    override fun toString(): String {
        return path
    }
}
